### ISSUE TITLE
Cannot get this package to work in most major cloud providers

### ISSUE THREAD

**Issue by jordanvector** [newrelic/newrelic-node-nextjs#135]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
 I have configured new relic in my NextJs application, it works from my local (can see metrics on the dashboard), but when I deploy the application on AWS Amplify, Vercel, or Azure Static web apps, I am not getting any metrics.

Currently looks like Amplify is not supported - https://github.com/newrelic/newrelic-node-nextjs/issues/91
Neither is Vercel - https://github.com/newrelic/newrelic-node-nextjs/issues/94

I have also deployed to azure static web apps, and also does not work there either.

Can someone please on the product team advise on what cloud providers are actually supported, currently it seems to only work locally.

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT share sensitive information, whether personal, proprietary, or otherwise! )

Not getting any APM data when deployed to cloud providers

## Expected Behavior

[NOTE]: # ( Tell us what you expected to happen. )

Expect to see APM data of applications deployed in cloud providers

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results

[NOTE]: # ( Provide any other relevant log data. )
[TIP]:  # ( Scrub logs and diagnostic information for sensitive information )

## Steps to Reproduce

I have followed the "Frontend Observability" section of this https://newrelic.com/blog/how-to-relic/nextjs-monitor-application-data for setting up client-side monitoring

[NOTE]: # ( Please be as specific as possible. )
[TIP]:  # ( Link a sample application that demonstrates the issue. )

**Comment by bizob2828** [comment]
@jordanvector you're correct in your assessment. We have on our near term roadmap items to solve working in Vercel and Amplify. The big issue is figuring out how to control the start of a next app. We will keep you posted.

**Comment by luizcarlosrodrigues** [comment]
Same issue here. looking foward for this feature

**Comment by luizcarlosrodrigues** [comment]
Does anyone have a timeline for this feature?

**Comment by jordanvector** [comment]
I got this to work in azure static web apps, I am not sure if this is the correct way but I will share for others.  Note: this will not work for vercel, because vercel has the issue where you cannot set independent variables between build and runtime.

I think the main issue for me in azure was this:

1. swa essentially due to their limitations forces you to use standalone mode. 
2. in standalone mode, since @newrelic/next is not including within the app anywhere, it is not added via the build therefore does not work.

What I did was this:

I included the @newrelic/next package via the new instrumentation hook defined here: https://nextjs.org/docs/pages/building-your-application/optimizing/instrumentation

Basically you set up a file called instrumentation.js in the root of the project, mine looks like this:

```
export async function register() {
    if (process.env.NEXT_RUNTIME === 'nodejs') {
        await import('@newrelic/next')
    }
}
```

In next.config.js add 

```
experimental: {instrumentationHook: true}
```

Start script must still contain NODE_OPTIONS='-r @newrelic/next'

After adding this and deploying I started getting proper APM as well as browser data. I actually had to alter the official example for browser data as well if anyone wants to see that code.

**Comment by luizcarlosrodrigues** [comment]
Anyone have tested what @jordanvector did on azure on AWS amplify?

**Comment by marcmarcet** [comment]
@jordanvector  can you share the snippet to get browser data working?

**Comment by jordanvector** [comment]
@marcmarcet yep. Here is our _document.js

```import newrelic from 'newrelic'
import dynamic from 'next/dynamic'
import Script from 'next/script'
import Document from 'next/document'
import {theme} from '../theme/index'
import createEmotionServer from '@emotion/server/create-instance'
import {createEmotionCache} from '../utils/create-emotion-cache'
const Html = dynamic(() => import('next/document').then((mod) => mod.Html))
const Head = dynamic(() => import('next/document').then((mod) => mod.Head))
const Main = dynamic(() => import('next/document').then((mod) => mod.Main))
const NextScript = dynamic(() => import('next/document').then((mod) => mod.NextScript))
class MyDocument extends Document {

  render() {
    return (
      <Html lang="en">
        <Head>
          <meta
            name="application-name"
            content={process.env.NEXT_PUBLIC_APP_NAME}
          />
          <meta
            name='format-detection'
            content='telephone=no'
          />
          <meta
            name="theme-color"
            content={theme.palette.primary.main}
          />
          <meta
            name="msapplication-navbutton-color"
            content={theme.palette.primary.main}
          />
          <meta
            name="apple-mobile-web-app-capable"
            content="yes"
          />
          <meta
            name='mobile-web-app-capable'
            content='yes'
          />
          <meta
            name="apple-mobile-web-app-status-bar-style"
            content={theme.palette.primary.main}
          />
          <meta
            httpEquiv="X-UA-Compatible"
            content="IE=edge"
          />
          <link
            rel="manifest"
            href="/manifest.json"
            crossOrigin="use-credentials"
          />
          <script
            type="text/javascript"
            dangerouslySetInnerHTML={{ __html: this.props.browserTimingHeader }}
          />
          <meta
            name='emotion-insertion-point'
            content=''
          />
          {this.props.emotionStyleTags}
        </Head>
        <body>
        {process.env.NEXT_PUBLIC_URL !== "http://localhost:3000" ? (
          <Script
              id="google-tag-manager"
              strategy='lazyOnload'
              dangerouslySetInnerHTML={{
                __html: `
                  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
                  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
                  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
                  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
                  })(window,document,'script','dataLayer', '${process.env.NEXT_PUBLIC_GTM_ID}');
                `,
              }}
            />
        ):null} 
        {process.env.NEXT_PUBLIC_URL !== "http://localhost:3000" ? (
          <noscript>
            <iframe
              src={`https://www.googletagmanager.com/ns.html?id=${process.env.NEXT_PUBLIC_GTM_ID}`}
              height="0"
              width="0"
              style={{ display: 'none', visibility: 'hidden' }}
            />
        </noscript>
        ):null}
        <Main />
        <NextScript />
        </body>
      </Html>
    );
  }
}
MyDocument.getInitialProps = async (ctx) => {
  const originalRenderPage = ctx.renderPage;
  let browserTimingHeader = ''
  // You can consider sharing the same Emotion cache between all the SSR requests to speed up performance.
  const cache = createEmotionCache();
  const { extractCriticalToChunks } = createEmotionServer(cache);

  ctx.renderPage = () =>
    originalRenderPage({
      enhanceApp: (App) =>
        function EnhanceApp(props) {
          return (
            <App
              emotionCache={cache}
              {...props}
            />
          );
        },
    });

  const initialProps = await Document.getInitialProps(ctx);
  const emotionStyles = extractCriticalToChunks(initialProps.html);
  const emotionStyleTags = emotionStyles.styles.map((style) => (
    <style
      data-emotion={`${style.key} ${style.ids.join(' ')}`}
      key={style.key}
      // eslint-disable-next-line react/no-danger
      dangerouslySetInnerHTML={{ __html: style.css }}
    />
  ));
  /**
 * For SSG pages the build is faster than the agent connect cycle
 * In those cases, let's wait for the agent to connect before getting
 * the browser agent script.
 */
  if(ctx.req && process.env.NEXT_RUNTIME === 'nodejs'){
    if (!newrelic.agent.collector.isConnected()) {
      await new Promise((resolve) => {
        newrelic.agent.on("connected", resolve);
      });
    }
    if(newrelic?.getBrowserTimingHeader) {
      browserTimingHeader = newrelic.getBrowserTimingHeader({
        hasToRemoveScriptWrapper: true,
        allowTransactionlessInjection: true,
      });
    }
  }

  return {
    ...initialProps,
    emotionStyleTags,
    browserTimingHeader,
  };
};
export default MyDocument;
```

../utils/create-emotion-cache Looks like this


```
import createCache from '@emotion/cache';
const isBrowser = typeof document !== 'undefined';

export const createEmotionCache = () => {
  let insertionPoint;

  if (isBrowser) {
    const emotionInsertionPoint = document.querySelector('meta[name="emotion-insertion-point"]');
    insertionPoint = emotionInsertionPoint ?? undefined;
  }

  return createCache({ key: 'mui-style', insertionPoint, prepend: true });
};

```

**Comment by bizob2828** [comment]
> I got this to work in azure static web apps, I am not sure if this is the correct way but I will share for others. Note: this will not work for vercel, because vercel has the issue where you cannot set independent variables between build and runtime.
> 
> I think the main issue for me in azure was this:
> 
> 1. swa essentially due to their limitations forces you to use standalone mode.
> 2. in standalone mode, since @newrelic/next is not including within the app anywhere, it is not added via the build therefore does not work.
> 
> What I did was this:
> 
> I included the @newrelic/next package via the new instrumentation hook defined here: https://nextjs.org/docs/pages/building-your-application/optimizing/instrumentation
> 
> Basically you set up a file called instrumentation.js in the root of the project, mine looks like this:
> 
> ```
> export async function register() {
>     if (process.env.NEXT_RUNTIME === 'nodejs') {
>         await import('@newrelic/next')
>     }
> }
> ```
> 
> In next.config.js add
> 
> ```
> experimental: {instrumentationHook: true}
> ```
> 

@jordanvector This is something I've tested locally to work as an alternative way to load `@newrelic/next`. I plan on documenting this and updating our sample app.  I'm testing this out on the various cloud providers as we speak.  I don't think this will work in Vercel just due to how the given components of Next.js are running so I'm digging in there but this should be fine in Azure if you said it already works.

> Start script must still contain NODE_OPTIONS='-r @newrelic/next'

This is not necessary as it replicates what's in instrumentation.js

> 
> After adding this and deploying I started getting proper APM as well as browser data. I actually had to alter the official example for browser data as well if anyone wants to see that code.

**Comment by jordanvector** [comment]
Updated my code...in the 11.x version of new relic this section is no longer needed, which has been removed from the above example

    if (!newrelic.agent.collector.isConnected()) {
      await new Promise((resolve) => {
        newrelic.agent.on("connected", resolve);
      });
    }

**Comment by armanTakmazyan** [comment]
In my case, I also needed to utilize `dotenv` as described here:
https://github.com/newrelic/newrelic-node-nextjs/issues/136

**Comment by ygerass** [comment]
Cannot prioritize this issue at the moment. Closing now, may reconsider later should the need persist.

**Comment by ygerass** [comment]
Next.js has evolved since we added instrumentation. Most of the use cases are around deploying to cloud providers like Vercel, AWS Amplify, Netlify, Azure Static Sites, etc. Most of these providers would not allow to control the run phase of Next.js CLI which is where we recommend registering the Next.js instrumentation. That being said, our recommendation is to rely on the experimental instrumentationHook feature of Next.js and use OpenTelemetry and send the data to New Relic via OTLP ([gist reference](https://gist.github.com/bizob2828/a7d52ac69b7db6d8508c865aafa0b62b)).

#nr-dev
